helix-term/src/handlers/completion.rs

```diff
@@ -0,0 +1,412 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Duration;
+
+use arc_swap::ArcSwap;
+use futures_util::stream::FuturesUnordered;
+use helix_core::chars::char_is_word;
+use helix_core::syntax::LanguageServerFeature;
+use helix_event::{
+    cancelable_future, cancelation, register_hook, send_blocking, CancelRx, CancelTx,
+};
+use helix_lsp::lsp;
+use helix_lsp::util::pos_to_lsp_pos;
+use helix_view::document::{Mode, SavePoint};
+use helix_view::handlers::lsp::{CompletionEvent, CompletionTrigger};
+use helix_view::Editor;
+use tokio::sync::mpsc::Sender;
+use tokio::time::Instant;
+use tokio_stream::StreamExt;
+
+use crate::commands;
+use crate::compositor::Compositor;
+use crate::config::Config;
+use crate::events::{OnModeSwitch, PostCommand, PostInsertChar};
+use crate::handlers::rope_ends_with;
+use crate::job::{dispatch, dispatch_blocking};
+use crate::keymap::MappableCommand;
+use crate::ui::editor::InsertEvent;
+use crate::ui::lsp::SignatureHelp;
+use crate::ui::{self, CompletionItem, Popup};
+
+use super::Handlers;
+
+#[derive(Debug)]
+pub(super) struct CompletionHandler {
+    /// currently active trigger which will cause a
+    /// completion request after the timeout
+    trigger: Option<CompletionTrigger>,
+    /// A handle for currently active completion request.
+    /// This can be used to determine whether the current
+    /// request is still active (and new triggers should be
+    /// ignored) and can also be used to abort the current
+    /// request (by dropping the handle)
+    request: Option<CancelTx>,
+    config: Arc<ArcSwap<Config>>,
+}
+
+impl CompletionHandler {
+    pub fn new(config: Arc<ArcSwap<Config>>) -> CompletionHandler {
+        Self {
+            config,
+            request: None,
+            trigger: None,
+        }
+    }
+}
+
+impl helix_event::AsyncHook for CompletionHandler {
+    type Event = CompletionEvent;
+
+    fn handle_event(
+        &mut self,
+        event: Self::Event,
+        _old_timeout: Option<Instant>,
+    ) -> Option<Instant> {
+        match event {
+            CompletionEvent::Trigger(trigger) => {
+                self.trigger = Some(trigger);
+            }
+            CompletionEvent::Cancel => {
+                self.trigger = None;
+                self.request = None;
+            }
+            CompletionEvent::DeleteText { pos } => {
+                // if we deleted the original trigger, abort the completion
+                if matches!(self.trigger, Some(CompletionTrigger{ trigger_pos,..}) if trigger_pos < pos)
+                {
+                    self.trigger = None;
+                    self.request = None;
+                }
+            }
+            CompletionEvent::Manual => {
+                // immediately request completions and drop all auto completion requests
+                self.request = None;
+                self.trigger = None;
+                self.finish_debounce();
```

Comment:
I wonder if it would make sense to centralize the call to `finish_debounce`. As it stands, if we have an event that we want to handle right away, the hook itself invokes its own execution by calling `finish_debounce`, and only by returning `None` does it come to be that it is not called a second time back in `run`.

I wonder if instead if a hook wants execution immediately, it can just return `None` and then the `run` function takes care of calling `finish_debounce`.

Reply:
Changing the meaning of `None` would mean its not possible for an `AsyncHook` to "abort" a debounce (which we already use here quite a bit for example for cancel events and also deleting back before the original trigger).

I considered adding an explicit enum with a cancel, execute imminently and deounce with timeout  <X>. Untimely I found that less clear then working with the option (giving the varaints meaingful names was hard and you usually endup working with option anyway)